File: web/src/test/java/no/nav/sbl/dialogarena/rest/ressurser/inntekt/VerdiRessursTest.java
package no.nav.sbl.dialogarena.rest.ressurser.inntekt;

import no.nav.sbl.dialogarena.rest.ressurser.inntekt.VerdiRessurs.VerdierFrontend;

import no.nav.sbl.dialogarena.sendsoknad.domain.oidc.SubjectHandlerWrapper;
import no.nav.sbl.dialogarena.sikkerhet.Tilgangskontroll;
import no.nav.sbl.dialogarena.soknadinnsending.business.service.TextService;
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOkonomibekreftelse;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOkonomibeskrivelserAvAnnet;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.oversikt.JsonOkonomioversiktFormue;
import no.nav.sbl.sosialhjelp.domain.SoknadUnderArbeid;
import no.nav.sbl.sosialhjelp.soknadunderbehandling.SoknadUnderArbeidRepository;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.List;

import static java.util.Arrays.asList;

import static no.nav.sbl.dialogarena.soknadinnsending.business.service.soknadservice.SoknadService.createEmptyJsonInternalSoknad;
import static no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.*;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.nullValue;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class VerdiRessursTest {

    private static final String BEHANDLINGSID = "123";
    private static final String EIER = "123456789101";

    @Mock
    private SoknadUnderArbeidRepository soknadUnderArbeidRepository;

    @Mock
    private Tilgangskontroll tilgangskontroll;

    @Mock
    private TextService textService;

    @Mock
    private SubjectHandlerWrapper subjectHandlerWrapper;

    @InjectMocks
    private VerdiRessurs verdiRessurs;

    @Before
    public void setUp() {
        when(subjectHandlerWrapper.getIdent()).thenReturn("123");
        when(textService.getJsonOkonomiTittel(anyString())).thenReturn("tittel");
    }

    @Test
    public void getVerdierSkalReturnereBekreftelseLikNullOgAltFalse(){
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER)));

        VerdierFrontend verdierFrontend = verdiRessurs.hentVerdier(BEHANDLINGSID);

        assertThat(verdierFrontend.bekreftelse, nullValue());
        assertFalse(verdierFrontend.bolig);
        assertFalse(verdierFrontend.campingvogn);
        assertFalse(verdierFrontend.kjoretoy);
        assertFalse(verdierFrontend.fritidseiendom);
        assertFalse(verdierFrontend.annet);
        assertThat(verdierFrontend.beskrivelseAvAnnet, nullValue());
    }

    @Test
    public void getVerdierSkalReturnereBekreftelserLikTrue(){
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithVerdier(true, asList(VERDI_BOLIG, VERDI_CAMPINGVOGN, VERDI_KJORETOY,
                        VERDI_FRITIDSEIENDOM, VERDI_ANNET), null));

        VerdierFrontend verdierFrontend = verdiRessurs.hentVerdier(BEHANDLINGSID);

        assertTrue(verdierFrontend.bekreftelse);
        assertTrue(verdierFrontend.bolig);
        assertTrue(verdierFrontend.campingvogn);
        assertTrue(verdierFrontend.kjoretoy);
        assertTrue(verdierFrontend.fritidseiendom);
        assertTrue(verdierFrontend.annet);
        assertThat(verdierFrontend.beskrivelseAvAnnet, nullValue());
    }

    @Test
    public void getVerdierSkalReturnereBeskrivelseAvAnnet(){
        String beskrivelse = "Bestefars klokke";
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithVerdier(true, asList(VERDI_ANNET), beskrivelse));

        VerdierFrontend verdierFrontend = verdiRessurs.hentVerdier(BEHANDLINGSID);

        assertTrue(verdierFrontend.bekreftelse);
        assertTrue(verdierFrontend.annet);
        assertThat(verdierFrontend.beskrivelseAvAnnet, is(beskrivelse));
    }

    @Test
    public void putVerdierSkalSetteAltFalseDersomManVelgerHarIkkeVerdier(){
        doNothing().when(tilgangskontroll).verifiserAtBrukerKanEndreSoknad(anyString());
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithVerdier(true, asList(VERDI_BOLIG, VERDI_CAMPINGVOGN, VERDI_KJORETOY,
                        VERDI_ANNET), "Bestefars klokke"));

        VerdierFrontend verdierFrontend = new VerdierFrontend();
        verdierFrontend.setBekreftelse(false);
        verdiRessurs.updateVerdier(BEHANDLINGSID, verdierFrontend);

        SoknadUnderArbeid soknadUnderArbeid = catchSoknadUnderArbeidSentToOppdaterSoknadsdata();
        List<JsonOkonomibekreftelse> bekreftelser = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData()
                .getOkonomi().getOpplysninger().getBekreftelse();
        JsonOkonomibekreftelse verdiBekreftelse = bekreftelser.get(0);
        List<JsonOkonomioversiktFormue> verdier = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData()
                .getOkonomi().getOversikt().getFormue();
        assertFalse(verdiBekreftelse.getVerdi());
        assertTrue(verdier.isEmpty());
    }

    @Test
    public void putVerdierSkalSetteAlleBekreftelserLikFalse(){
        doNothing().when(tilgangskontroll).verifiserAtBrukerKanEndreSoknad(anyString());
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithVerdier(true, asList(VERDI_BOLIG, VERDI_CAMPINGVOGN,
                        VERDI_KJORETOY, VERDI_ANNET), "Bestefars klokke"));

        VerdierFrontend verdierFrontend = new VerdierFrontend();
        verdierFrontend.setBekreftelse(false);
        verdiRessurs.updateVerdier(BEHANDLINGSID, verdierFrontend);

        SoknadUnderArbeid soknadUnderArbeid = catchSoknadUnderArbeidSentToOppdaterSoknadsdata();
        List<JsonOkonomibekreftelse> bekreftelser = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData()
                .getOkonomi().getOpplysninger().getBekreftelse();
        JsonOkonomibekreftelse verdiBekreftelse = bekreftelser.get(0);
        List<JsonOkonomioversiktFormue> verdier = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData()
                .getOkonomi().getOversikt().getFormue();
        String beskrivelse = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData()
                .getOkonomi().getOpplysninger().getBeskrivelseAvAnnet().getVerdi();
        assertFalse(verdiBekreftelse.getVerdi());
        assertTrue(verdier.isEmpty());
        assertThat(beskrivelse, is(""));
    }

    @Test
    public void putVerdierSkalSetteNoenBekreftelser(){
        doNothing().when(tilgangskontroll).verifiserAtBrukerKanEndreSoknad(anyString());
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER)));

        VerdierFrontend verdierFrontend = new VerdierFrontend();
        verdierFrontend.setBekreftelse(true);
        verdierFrontend.setBolig(true);
        verdierFrontend.setCampingvogn(true);
        verdierFrontend.setFritidseiendom(false);
        verdierFrontend.setKjoretoy(false);
        verdierFrontend.setAnnet(false);
        verdiRessurs.updateVerdier(BEHANDLINGSID, verdierFrontend);

        SoknadUnderArbeid soknadUnderArbeid = catchSoknadUnderArbeidSentToOppdaterSoknadsdata();
        List<JsonOkonomibekreftelse> bekreftelser = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData()
                .getOkonomi().getOpplysninger().getBekreftelse();
        JsonOkonomibekreftelse verdiBekreftelse = bekreftelser.get(0);
        List<JsonOkonomioversiktFormue> verdier = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData()
                .getOkonomi().getOversikt().getFormue();
        assertThat(verdiBekreftelse.getKilde(), is(JsonKilde.BRUKER));
        assertThat(verdiBekreftelse.getType(), is(BEKREFTELSE_VERDI));
        assertTrue(verdiBekreftelse.getVerdi());
        assertTrue(verdier.stream().anyMatch(verdi -> verdi.getType().equals(VERDI_BOLIG)));
        assertTrue(verdier.stream().anyMatch(verdi -> verdi.getType().equals(VERDI_CAMPINGVOGN)));
        assertFalse(verdier.stream().anyMatch(verdi -> verdi.getType().equals(VERDI_KJORETOY)));
        assertFalse(verdier.stream().anyMatch(verdi -> verdi.getType().equals(VERDI_FRITIDSEIENDOM)));
        assertFalse(verdier.stream().anyMatch(verdi -> verdi.getType().equals(VERDI_ANNET)));
    }

    @Test
    public void putVerdierSkalSetteAlleBekreftelser(){
        doNothing().when(tilgangskontroll).verifiserAtBrukerKanEndreSoknad(anyString());
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER)));

        VerdierFrontend verdierFrontend = new VerdierFrontend();
        verdierFrontend.setBekreftelse(true);
        verdierFrontend.setBolig(true);
        verdierFrontend.setCampingvogn(true);
        verdierFrontend.setFritidseiendom(true);
        verdierFrontend.setKjoretoy(true);
        verdierFrontend.setAnnet(true);
        verdiRessurs.updateVerdier(BEHANDLINGSID, verdierFrontend);

        SoknadUnderArbeid soknadUnderArbeid = catchSoknadUnderArbeidSentToOppdaterSoknadsdata();
        List<JsonOkonomibekreftelse> bekreftelser = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData()
                .getOkonomi().getOpplysninger().getBekreftelse();
        JsonOkonomibekreftelse verdiBekreftelse = bekreftelser.get(0);
        List<JsonOkonomioversiktFormue> verdier = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData()
                .getOkonomi().getOversikt().getFormue();
        assertThat(verdiBekreftelse.getKilde(), is(JsonKilde.BRUKER));
        assertThat(verdiBekreftelse.getType(), is(BEKREFTELSE_VERDI));
        assertTrue(verdiBekreftelse.getVerdi());
        assertTrue(verdier.stream().anyMatch(verdi -> verdi.getType().equals(VERDI_KJORETOY)));
        assertTrue(verdier.stream().anyMatch(verdi -> verdi.getType().equals(VERDI_CAMPINGVOGN)));
        assertTrue(verdier.stream().anyMatch(verdi -> verdi.getType().equals(VERDI_BOLIG)));
        assertTrue(verdier.stream().anyMatch(verdi -> verdi.getType().equals(VERDI_FRITIDSEIENDOM)));
        assertTrue(verdier.stream().anyMatch(verdi -> verdi.getType().equals(VERDI_ANNET)));
    }

    @Test
    public void putVerdierSkalFjerneBeskrivelseAvAnnetDersomAnnetBlirAvkreftet(){
        doNothing().when(tilgangskontroll).verifiserAtBrukerKanEndreSoknad(anyString());
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithVerdier(true, asList(VERDI_ANNET), "Vinylplater"));

        VerdierFrontend verdierFrontend = new VerdierFrontend();
        verdierFrontend.setBekreftelse(false);
        verdiRessurs.updateVerdier(BEHANDLINGSID, verdierFrontend);

        SoknadUnderArbeid soknadUnderArbeid = catchSoknadUnderArbeidSentToOppdaterSoknadsdata();
        List<JsonOkonomibekreftelse> bekreftelser = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData()
                .getOkonomi().getOpplysninger().getBekreftelse();
        JsonOkonomibekreftelse verdiBekreftelse = bekreftelser.get(0);
        String beskrivelse = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData()
                .getOkonomi().getOpplysninger().getBeskrivelseAvAnnet().getVerdi();
        assertFalse(verdiBekreftelse.getVerdi());
        assertThat(beskrivelse, is(""));
    }

    private SoknadUnderArbeid catchSoknadUnderArbeidSentToOppdaterSoknadsdata() {
        ArgumentCaptor<SoknadUnderArbeid> argument = ArgumentCaptor.forClass(SoknadUnderArbeid.class);
        verify(soknadUnderArbeidRepository).oppdaterSoknadsdata(argument.capture(), anyString());
        return argument.getValue();
    }

    private SoknadUnderArbeid createJsonInternalSoknadWithVerdier(Boolean harVerdier, List<String> verdiTyper, String beskrivelseAvAnnet) {
        SoknadUnderArbeid soknadUnderArbeid = new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER));
        List<JsonOkonomioversiktFormue> verdier = new ArrayList<>();
        for (String verdi: verdiTyper) {
            verdier.add(new JsonOkonomioversiktFormue()
                    .withKilde(JsonKilde.BRUKER)
                    .withType(verdi)
                    .withTittel("tittel"));
        }
        soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getOkonomi().getOpplysninger().setBekreftelse(asList(new JsonOkonomibekreftelse()
                .withKilde(JsonKilde.BRUKER)
                .withType(BEKREFTELSE_VERDI)
                .withVerdi(harVerdier)));
        soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getOkonomi().getOversikt().setFormue(verdier);
        soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getOkonomi().getOpplysninger().setBeskrivelseAvAnnet(
                new JsonOkonomibeskrivelserAvAnnet().withVerdi(beskrivelseAvAnnet));
        return soknadUnderArbeid;
    }
}

File: src/test/kotlin/no/nav/sosialhjelp/soknad/inntekt/verdi/VerdiRessursTest.kt
package no.nav.sosialhjelp.soknad.inntekt.verdi

import io.mockk.clearAllMocks
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.runs
import io.mockk.slot
import io.mockk.verify
import no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper
import no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.VERDI_ANNET
import no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.VERDI_BOLIG
import no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.VERDI_CAMPINGVOGN
import no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.VERDI_FRITIDSEIENDOM
import no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.VERDI_KJORETOY
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOkonomibekreftelse
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOkonomibeskrivelserAvAnnet
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.oversikt.JsonOkonomioversiktFormue
import no.nav.sosialhjelp.soknad.business.db.repositories.soknadunderarbeid.SoknadUnderArbeidRepository
import no.nav.sosialhjelp.soknad.business.service.TextService
import no.nav.sosialhjelp.soknad.common.subjecthandler.StaticSubjectHandlerImpl
import no.nav.sosialhjelp.soknad.common.subjecthandler.SubjectHandlerUtils
import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeid
import no.nav.sosialhjelp.soknad.domain.model.exception.AuthorizationException
import no.nav.sosialhjelp.soknad.innsending.SoknadService.Companion.createEmptyJsonInternalSoknad
import no.nav.sosialhjelp.soknad.inntekt.verdi.VerdiRessurs.VerdierFrontend
import no.nav.sosialhjelp.soknad.tilgangskontroll.Tilgangskontroll
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class VerdiRessursTest {

    private val BEHANDLINGSID = "123"
    private val EIER = "123456789101"

    private val soknadUnderArbeidRepository: SoknadUnderArbeidRepository = mockk()
    private val tilgangskontroll: Tilgangskontroll = mockk()
    private val textService: TextService = mockk()
    private val verdiRessurs = VerdiRessurs(tilgangskontroll, soknadUnderArbeidRepository, textService)

    @BeforeEach
    fun setUp() {
        clearAllMocks()
        System.setProperty("environment.name", "test")
        SubjectHandlerUtils.setNewSubjectHandlerImpl(StaticSubjectHandlerImpl())
    }

    @AfterEach
    fun tearDown() {
        SubjectHandlerUtils.resetSubjectHandlerImpl()
        System.clearProperty("environment.name")
    }

    @Test
    fun getVerdierSkalReturnereBekreftelseLikNullOgAltFalse() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER))

        val verdierFrontend = verdiRessurs.hentVerdier(BEHANDLINGSID)
        assertThat(verdierFrontend.bekreftelse).isNull()
        assertThat(verdierFrontend.bolig).isFalse
        assertThat(verdierFrontend.campingvogn).isFalse
        assertThat(verdierFrontend.kjoretoy).isFalse
        assertThat(verdierFrontend.fritidseiendom).isFalse
        assertThat(verdierFrontend.annet).isFalse
        assertThat(verdierFrontend.beskrivelseAvAnnet).isNull()
    }

    @Test
    fun getVerdierSkalReturnereBekreftelserLikTrue() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            createJsonInternalSoknadWithVerdier(
                true,
                listOf(VERDI_BOLIG, VERDI_CAMPINGVOGN, VERDI_KJORETOY, VERDI_FRITIDSEIENDOM, VERDI_ANNET),
                null
            )

        val verdierFrontend = verdiRessurs.hentVerdier(BEHANDLINGSID)
        assertThat(verdierFrontend.bekreftelse).isTrue
        assertThat(verdierFrontend.bolig).isTrue
        assertThat(verdierFrontend.campingvogn).isTrue
        assertThat(verdierFrontend.kjoretoy).isTrue
        assertThat(verdierFrontend.fritidseiendom).isTrue
        assertThat(verdierFrontend.annet).isTrue
        assertThat(verdierFrontend.beskrivelseAvAnnet).isNull()
    }

    @Test
    fun getVerdierSkalReturnereBeskrivelseAvAnnet() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } just runs
        val beskrivelse = "Bestefars klokke"
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            createJsonInternalSoknadWithVerdier(true, listOf(VERDI_ANNET), beskrivelse)

        val verdierFrontend: VerdierFrontend = verdiRessurs.hentVerdier(BEHANDLINGSID)
        assertThat(verdierFrontend.bekreftelse).isTrue
        assertThat(verdierFrontend.annet).isTrue
        assertThat(verdierFrontend.beskrivelseAvAnnet).isEqualTo(beskrivelse)
    }

    @Test
    fun putVerdierSkalSetteAltFalseDersomManVelgerHarIkkeVerdier() {
        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(any()) } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            createJsonInternalSoknadWithVerdier(
                true,
                listOf(VERDI_BOLIG, VERDI_CAMPINGVOGN, VERDI_KJORETOY, VERDI_ANNET),
                "Bestefars klokke"
            )
        every { textService.getJsonOkonomiTittel(any()) } returns "tittel"

        val slot = slot<SoknadUnderArbeid>()
        every { soknadUnderArbeidRepository.oppdaterSoknadsdata(capture(slot), any()) } just runs

        val verdierFrontend = VerdierFrontend(bekreftelse = false)
        verdiRessurs.updateVerdier(BEHANDLINGSID, verdierFrontend)

        val soknadUnderArbeid = slot.captured
        val bekreftelser = soknadUnderArbeid.jsonInternalSoknad.soknad.data.okonomi.opplysninger.bekreftelse
        val verdiBekreftelse = bekreftelser[0]
        val verdier = soknadUnderArbeid.jsonInternalSoknad.soknad.data.okonomi.oversikt.formue
        assertThat(verdiBekreftelse.verdi).isFalse
        assertThat(verdier.isEmpty()).isTrue
    }

    @Test
    fun putVerdierSkalSetteAlleBekreftelserLikFalse() {
        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(any()) } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            createJsonInternalSoknadWithVerdier(
                true,
                listOf(VERDI_BOLIG, VERDI_CAMPINGVOGN, VERDI_KJORETOY, VERDI_ANNET),
                "Bestefars klokke"
            )
        every { textService.getJsonOkonomiTittel(any()) } returns "tittel"

        val slot = slot<SoknadUnderArbeid>()
        every { soknadUnderArbeidRepository.oppdaterSoknadsdata(capture(slot), any()) } just runs

        val verdierFrontend = VerdierFrontend(bekreftelse = false)
        verdiRessurs.updateVerdier(BEHANDLINGSID, verdierFrontend)

        val soknadUnderArbeid = slot.captured
        val bekreftelser = soknadUnderArbeid.jsonInternalSoknad.soknad.data.okonomi.opplysninger.bekreftelse
        val verdiBekreftelse = bekreftelser[0]
        val verdier = soknadUnderArbeid.jsonInternalSoknad.soknad.data.okonomi.oversikt.formue
        val beskrivelse = soknadUnderArbeid.jsonInternalSoknad.soknad.data.okonomi.opplysninger.beskrivelseAvAnnet.verdi
        assertThat(verdiBekreftelse.verdi).isFalse
        assertThat(verdier.isEmpty()).isTrue
        assertThat(beskrivelse).isBlank
    }

    @Test
    fun putVerdierSkalSetteNoenBekreftelser() {
        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(any()) } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER))
        every { textService.getJsonOkonomiTittel(any()) } returns "tittel"

        val slot = slot<SoknadUnderArbeid>()
        every { soknadUnderArbeidRepository.oppdaterSoknadsdata(capture(slot), any()) } just runs

        val verdierFrontend = VerdierFrontend(
            bekreftelse = true,
            bolig = true,
            campingvogn = true,
            kjoretoy = false,
            fritidseiendom = false,
            annet = false
        )
        verdiRessurs.updateVerdier(BEHANDLINGSID, verdierFrontend)

        val soknadUnderArbeid = slot.captured
        val bekreftelser = soknadUnderArbeid.jsonInternalSoknad.soknad.data.okonomi.opplysninger.bekreftelse
        val verdiBekreftelse = bekreftelser[0]
        val verdier = soknadUnderArbeid.jsonInternalSoknad.soknad.data.okonomi.oversikt.formue
        assertThat(verdiBekreftelse.kilde).isEqualTo(JsonKilde.BRUKER)
        assertThat(verdiBekreftelse.type).isEqualTo(SoknadJsonTyper.BEKREFTELSE_VERDI)
        assertThat(verdiBekreftelse.verdi).isTrue
        assertThat(verdier.any { it.type == VERDI_BOLIG }).isTrue
        assertThat(verdier.any { it.type == VERDI_CAMPINGVOGN }).isTrue
        assertThat(verdier.any { it.type == VERDI_KJORETOY }).isFalse
        assertThat(verdier.any { it.type == VERDI_FRITIDSEIENDOM }).isFalse
        assertThat(verdier.any { it.type == VERDI_ANNET }).isFalse
    }

    @Test
    fun putVerdierSkalSetteAlleBekreftelser() {
        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(any()) } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER))
        every { textService.getJsonOkonomiTittel(any()) } returns "tittel"

        val slot = slot<SoknadUnderArbeid>()
        every { soknadUnderArbeidRepository.oppdaterSoknadsdata(capture(slot), any()) } just runs

        val verdierFrontend = VerdierFrontend(
            bekreftelse = true,
            bolig = true,
            campingvogn = true,
            kjoretoy = true,
            fritidseiendom = true,
            annet = true
        )
        verdiRessurs.updateVerdier(BEHANDLINGSID, verdierFrontend)

        val soknadUnderArbeid = slot.captured
        val bekreftelser = soknadUnderArbeid.jsonInternalSoknad.soknad.data.okonomi.opplysninger.bekreftelse
        val verdiBekreftelse = bekreftelser[0]
        val verdier = soknadUnderArbeid.jsonInternalSoknad.soknad.data.okonomi.oversikt.formue
        assertThat(verdiBekreftelse.kilde).isEqualTo(JsonKilde.BRUKER)
        assertThat(verdiBekreftelse.type).isEqualTo(SoknadJsonTyper.BEKREFTELSE_VERDI)
        assertThat(verdiBekreftelse.verdi).isTrue
        assertThat(verdier.any { it.type == VERDI_KJORETOY }).isTrue
        assertThat(verdier.any { it.type == VERDI_CAMPINGVOGN }).isTrue
        assertThat(verdier.any { it.type == VERDI_BOLIG }).isTrue
        assertThat(verdier.any { it.type == VERDI_FRITIDSEIENDOM }).isTrue
        assertThat(verdier.any { it.type == VERDI_ANNET }).isTrue
    }

    @Test
    fun putVerdierSkalFjerneBeskrivelseAvAnnetDersomAnnetBlirAvkreftet() {
        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(any()) } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            createJsonInternalSoknadWithVerdier(true, listOf(VERDI_ANNET), "Vinylplater")
        every { textService.getJsonOkonomiTittel(any()) } returns "tittel"

        val slot = slot<SoknadUnderArbeid>()
        every { soknadUnderArbeidRepository.oppdaterSoknadsdata(capture(slot), any()) } just runs

        val verdierFrontend = VerdierFrontend(bekreftelse = false)
        verdiRessurs.updateVerdier(BEHANDLINGSID, verdierFrontend)

        val soknadUnderArbeid = slot.captured
        val bekreftelser = soknadUnderArbeid.jsonInternalSoknad.soknad.data.okonomi.opplysninger.bekreftelse
        val verdiBekreftelse = bekreftelser[0]
        val beskrivelse = soknadUnderArbeid.jsonInternalSoknad.soknad.data.okonomi.opplysninger.beskrivelseAvAnnet.verdi
        assertThat(verdiBekreftelse.verdi).isFalse
        assertThat(beskrivelse).isBlank
    }

    @Test
    fun getVerdierSkalKasteAuthorizationExceptionVedManglendeTilgang() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } throws AuthorizationException("Not for you my friend")

        assertThatExceptionOfType(AuthorizationException::class.java)
            .isThrownBy { verdiRessurs.hentVerdier(BEHANDLINGSID) }

        verify(exactly = 0) { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) }
    }

    @Test
    fun putVerdierSkalKasteAuthorizationExceptionVedManglendeTilgang() {
        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(BEHANDLINGSID) } throws AuthorizationException("Not for you my friend")
        val verdierFrontend = VerdierFrontend()
        assertThatExceptionOfType(AuthorizationException::class.java)
            .isThrownBy { verdiRessurs.updateVerdier(BEHANDLINGSID, verdierFrontend) }

        verify(exactly = 0) { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) }
    }

    private fun createJsonInternalSoknadWithVerdier(
        harVerdier: Boolean,
        verdiTyper: List<String>,
        beskrivelseAvAnnet: String?
    ): SoknadUnderArbeid {
        val soknadUnderArbeid = SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER))
        val verdier: MutableList<JsonOkonomioversiktFormue> = ArrayList()
        for (verdi in verdiTyper) {
            verdier.add(
                JsonOkonomioversiktFormue()
                    .withKilde(JsonKilde.BRUKER)
                    .withType(verdi)
                    .withTittel("tittel")
            )
        }
        soknadUnderArbeid.jsonInternalSoknad.soknad.data.okonomi.opplysninger.bekreftelse = listOf(
            JsonOkonomibekreftelse()
                .withKilde(JsonKilde.BRUKER)
                .withType(SoknadJsonTyper.BEKREFTELSE_VERDI)
                .withVerdi(harVerdier)
        )
        soknadUnderArbeid.jsonInternalSoknad.soknad.data.okonomi.oversikt.formue = verdier
        soknadUnderArbeid.jsonInternalSoknad.soknad.data.okonomi.opplysninger.beskrivelseAvAnnet =
            JsonOkonomibeskrivelserAvAnnet().withVerdi(beskrivelseAvAnnet)
        return soknadUnderArbeid
    }
}
